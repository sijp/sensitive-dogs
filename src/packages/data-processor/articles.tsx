import lodash from "lodash";
import { ArticleContent } from "@sensitive-dogs/google-drive";

interface ArticleElement {
  text: string | null | undefined;
  italic: boolean | null | undefined;
  underline: boolean | null | undefined;
  bold: boolean | null | undefined;
  link: string | null | undefined;
  image: string | null | undefined;
  floatImages: string[] | null | undefined;
  horizontalLine: boolean | null | undefined;
  interactive?: string | null | undefined;
}

interface RawParagraph {
  type: string | null | undefined;
  elements: ArticleElement[] | undefined;
  bulletType: string | undefined;
  bullet: string | null | undefined;
}

interface Paragraph {
  type: string | null | undefined;
  elements: ArticleElement[] | undefined;
}

interface IndexedParagraph extends Paragraph {
  index: number;
}

interface List {
  type: "LIST";
  bulletType: string;
  paragraphs: Paragraph[];
}

interface IndexedList extends List {
  index: number;
}

type ParsedDocument = Array<Paragraph | List> | null;

function maybeConvertToInteractive(element: ArticleElement) {
  const youtube = (element.text || "").match(
    /https:\/\/www\.youtube\.com\/watch\?v=([\w-]+)/
  );
  if (youtube) {
    return {
      ...element,
      link: `https://www.youtube.com/embed/${youtube[1]}`,
      interactive: "youtube"
    };
  }
  return element;
}

function groupBulletedLists(paragraphs: RawParagraph[]) {
  const cleanedParagraphs = paragraphs.map((paragraph, index) => ({
    ...paragraph,
    index
  }));
  const listGroups = lodash.groupBy(
    cleanedParagraphs,
    (paragraph) => paragraph.bullet || paragraph.index
  );

  const indexedGroups = Object.entries(listGroups).map(
    ([index, paragraphs]) => {
      if (paragraphs.length === 1) {
        return [index, paragraphs[0]] as [string, IndexedParagraph];
      }
      return [
        index,
        {
          index: 0,
          type: "LIST",
          bulletType: paragraphs[0].bulletType || "ul",
          paragraphs: paragraphs.map(({ index, bullet, bulletType, ...p }) => p)
        }
      ] as [string, IndexedList];
    }
  );

  const tree = indexedGroups.reduce<
    Record<string, IndexedList | IndexedParagraph>
  >((memo, [index, group]) => {
    return { ...memo, [index]: group };
  }, {});

  const ret = lodash
    .uniqBy(
      cleanedParagraphs,
      (paragraph) => paragraph.bullet || paragraph.index
    )
    .map(({ bullet, bulletType, ...paragraph }) => {
      if (!bullet) return paragraph as Paragraph;
      const { index, ...list } = tree[bullet] as IndexedList;
      return list as List;
    });

  if ("kix.ovswf3ioox54" in listGroups) console.log(ret);

  return ret;
}

export function parseDocument(
  doc: ArticleContent,
  imagesIdMapper: { [key: string]: string }
): ParsedDocument {
  const { lists = {} } = doc;
  if (!doc.body?.content) return null;

  const paragraphs = doc.body.content.map(({ paragraph }) => {
    if (paragraph) {
      const {
        elements,
        paragraphStyle,
        bullet = {},
        positionedObjectIds
      } = paragraph;
      const listProperties = bullet.listId
        ? lists?.[bullet.listId].listProperties?.nestingLevels?.[0]
        : undefined;

      return {
        type: paragraphStyle?.namedStyleType,
        bullet: bullet.listId,
        bulletType: listProperties?.glyphSymbol
          ? "ul"
          : listProperties?.glyphType
          ? "ol"
          : undefined,
        elements: elements?.map(
          ({
            textRun = { textStyle: {} },
            inlineObjectElement = { textStyle: {} },
            horizontalRule
          }) => {
            const { content, textStyle = {} } = textRun;
            const { italic, underline, bold, link: textLink } = textStyle;
            const { inlineObjectId, textStyle: inlineTextStyle } =
              inlineObjectElement;
            const { link: imageLink } = inlineTextStyle || {};

            return maybeConvertToInteractive({
              text: content,
              italic,
              underline,
              bold,
              link: (textLink || imageLink)?.url,
              image: imagesIdMapper[inlineObjectId || ""],
              floatImages: positionedObjectIds?.map((id) => imagesIdMapper[id]),
              horizontalLine: !!horizontalRule
            });
          }
        )
      };
    }
    return null;
  });

  return groupBulletedLists(lodash.compact(paragraphs));
}
