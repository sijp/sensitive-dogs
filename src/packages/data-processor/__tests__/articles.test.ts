import { parseDocument } from "../articles";
import googleDoc from "../__mocks__/google-doc.mock";

describe("data-processor", () => {
  describe("articles", () => {
    describe("parseDocument", () => {
      it("Sholuld return null for content-less doc", () => {
        expect(parseDocument({}, {})).toBeNull();
        expect(parseDocument({ body: {} }, {})).toBeNull();
        // @ts-ignore - docs are runtime inputs, so no static type safety
        expect(parseDocument({ body: { content: " zzz " } }, {})).toBeNull();
      });

      it("Should parse a document", () => {
        expect(
          parseDocument(googleDoc, {
            "kix.abcdefg": "image1",
            "kix.1234567890": "image2"
          })
        ).toMatchSnapshot();
      });
    });
  });
});
