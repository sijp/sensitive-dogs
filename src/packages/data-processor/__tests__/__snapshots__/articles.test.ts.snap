// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`data-processor articles parseDocument Should parse a document 1`] = `
[
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "Title 1
",
        "underline": undefined,
      },
    ],
    "index": 0,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "Title 2
",
        "underline": undefined,
      },
    ],
    "index": 1,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "Title 3
",
        "underline": undefined,
      },
    ],
    "index": 2,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "Title 4
",
        "underline": undefined,
      },
    ],
    "index": 3,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "Title 5
",
        "underline": undefined,
      },
    ],
    "index": 4,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "Normal Paragraph
",
        "underline": undefined,
      },
    ],
    "index": 5,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "
",
        "underline": undefined,
      },
    ],
    "index": 6,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "A Paragraph with ",
        "underline": undefined,
      },
      {
        "bold": true,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "bold",
        "underline": undefined,
      },
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": ", ",
        "underline": undefined,
      },
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": true,
        "link": undefined,
        "text": "italic",
        "underline": undefined,
      },
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": " and ",
        "underline": undefined,
      },
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "underline",
        "underline": true,
      },
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": ". And all ",
        "underline": undefined,
      },
      {
        "bold": true,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": true,
        "link": undefined,
        "text": "together",
        "underline": true,
      },
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": ".
",
        "underline": undefined,
      },
    ],
    "index": 7,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "
",
        "underline": undefined,
      },
    ],
    "index": 8,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": "image1",
        "italic": undefined,
        "link": undefined,
        "text": undefined,
        "underline": undefined,
      },
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "
",
        "underline": undefined,
      },
    ],
    "index": 9,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "
",
        "underline": undefined,
      },
    ],
    "index": 10,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "Ordered Bullets:
",
        "underline": undefined,
      },
    ],
    "index": 11,
    "type": "NORMAL_TEXT",
  },
  {
    "bulletType": "ol",
    "paragraphs": [
      {
        "elements": [
          {
            "bold": undefined,
            "floatImages": undefined,
            "horizontalLine": false,
            "image": undefined,
            "italic": undefined,
            "link": undefined,
            "text": "One
",
            "underline": undefined,
          },
        ],
        "type": "NORMAL_TEXT",
      },
      {
        "elements": [
          {
            "bold": undefined,
            "floatImages": undefined,
            "horizontalLine": false,
            "image": undefined,
            "italic": undefined,
            "link": undefined,
            "text": "Two
",
            "underline": undefined,
          },
        ],
        "type": "NORMAL_TEXT",
      },
      {
        "elements": [
          {
            "bold": undefined,
            "floatImages": undefined,
            "horizontalLine": false,
            "image": undefined,
            "italic": undefined,
            "link": undefined,
            "text": "Three
",
            "underline": undefined,
          },
        ],
        "type": "NORMAL_TEXT",
      },
    ],
    "type": "LIST",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "
",
        "underline": undefined,
      },
    ],
    "index": 15,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "Unordered Bullets:
",
        "underline": undefined,
      },
    ],
    "index": 16,
    "type": "NORMAL_TEXT",
  },
  {
    "bulletType": "ul",
    "paragraphs": [
      {
        "elements": [
          {
            "bold": undefined,
            "floatImages": undefined,
            "horizontalLine": false,
            "image": undefined,
            "italic": undefined,
            "link": undefined,
            "text": "First
",
            "underline": undefined,
          },
        ],
        "type": "NORMAL_TEXT",
      },
      {
        "elements": [
          {
            "bold": undefined,
            "floatImages": undefined,
            "horizontalLine": false,
            "image": undefined,
            "italic": undefined,
            "link": undefined,
            "text": "Second
",
            "underline": undefined,
          },
        ],
        "type": "NORMAL_TEXT",
      },
      {
        "elements": [
          {
            "bold": undefined,
            "floatImages": undefined,
            "horizontalLine": false,
            "image": undefined,
            "italic": undefined,
            "link": undefined,
            "text": "Third
",
            "underline": undefined,
          },
        ],
        "type": "NORMAL_TEXT",
      },
    ],
    "type": "LIST",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "
",
        "underline": undefined,
      },
    ],
    "index": 20,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "
",
        "underline": undefined,
      },
    ],
    "index": 21,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "
",
        "underline": undefined,
      },
    ],
    "index": 22,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": [
          "image2",
        ],
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "This is a long paragraph with a floating image This is a long paragraph with a floating image This is a long paragraph with a floating image This is a long paragraph with a floating image This is a long paragraph with a floating image This is a long paragraph with a floating image This is a long paragraph with a floating image This is a long paragraph with a floating image.
",
        "underline": undefined,
      },
    ],
    "index": 23,
    "type": "NORMAL_TEXT",
  },
  {
    "elements": [
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "interactive": "youtube",
        "italic": undefined,
        "link": "https://www.youtube.com/embed/12345678901",
        "text": "https://www.youtube.com/watch?v=12345678901",
        "underline": true,
      },
      {
        "bold": undefined,
        "floatImages": undefined,
        "horizontalLine": false,
        "image": undefined,
        "italic": undefined,
        "link": undefined,
        "text": "
",
        "underline": undefined,
      },
    ],
    "index": 24,
    "type": "NORMAL_TEXT",
  },
]
`;
